### DISCUSSION TITLE
Loss of contact under acceleration

### DISCUSSION THREAD

**Discussion by stracovskylu** [google-deepmind/mujoco#111] [Asking for Help] [answered]
I've been trying to use mujoco for grasp simulations, and have noticed that under acceleration some contact distances go above zero and lose contact (even if the acceleration and contact normal are perpendicular). Increasing the contact normal force far above what what ideally would maintain contact reduces the frequency of the contact loss, but isn't desirable for my application. Is there any way to reduce or address this occurrence? I thought that adjusting the solref and impratio parameters might help, but so far the parameters I have tried did not seem to produce any significant changes.

I have attached some code (written using mujocopy) that demonstrates what I've described. (a .txt because I couldn't attach a .py)
[contact_demo.txt](https://github.com/deepmind/mujoco/files/7876002/contact_demo.txt)

**Comment by yuvaltassa** [comment]
Unfortunately we do not support mujoco-py. If you can attach a pure XML file where your issue can be observed using `simulate` that would be ideal. Alternatively, if you make a dm_control colab based on e.g. the [tutorial colab](https://github.com/deepmind/dm_control/blob/master/dm_control/mujoco/tutorial.ipynb), we could look at that.

**Reply by stracovskylu** [reply]
Thanks for getting back to me. I've made a [colab notebook](https://github.com/stracovskylu/misc_mujoco_test/blob/main/mujoco_contact_demo.ipynb) to demonstrate what I'm asking about.
In the notebook are three cases where objects are lifted and held via friction, which have varying stability.
- The first where a sphere is lifted seems reasonable to me. Oscillation of the contact distance only appears after the object undergoes acceleration, and disappears after the lift stops.
- The second case is contact on opposing edges of a cube, which has results similar to the first case, except the contact distance oscillation takes longer to disappear.
- The final case is contact on opposite sides of a cylinder, where even after the lift is finished there is continued contact distance oscillation.

Contact distance oscillation makes some sense to me given what I read in the mujoco documentation on the way contact is modeled, but I'm not entirely sure how I can make sure this behavior is relatively realistic. Presumably changing the solref, solimp, and impratio parameters could do this, but given what happens in the notebook it appears that geometry at the point of contact plays a significant factor. Would this mean I have to tune the parameters for pretty much every contact situation?

Please let me know if there's any issue accessing or running the notebook.

**Comment by yuvaltassa** [comment, marked as answer]
Thanks for the colab, I've taken a look.

Firstly, two recommendations.

1. Colab is not the right tool for this type of investigation. You really want the interactivity of something like `simulate` where you can interactively pull on objects, visualise contact forces / points and change contact parameters interactively (using the contact override mechanism). This will allow you to develop an intuition for the effects of different settings.
2. If you insist on using colab, at least set `scene_option.flags[enums.mjtVisFlag.mjVIS_CONTACTPOINT] = True` so you'll see the contact points.

Second, I could stabilise all three objects by reducing `solref[0]` to 5ms (=.005) (while reverting everything else to default) and increasing the grip force. 

You are dividing by gear when setting ctrl, why? It means gear has no effect (why not just set it to 1?) and you are using a grip force that is insufficient to hold to objects in place at high forces/accelerations (e.g. when hitting the top slider limit), so they slip. Remember that the thing that is bounded are impulses; forces are unbounded as objects are more and more rigid (though of course real objects are never infinitely rigid). E.g. when you gently tap a coin onto a glass window you are easily creating instantaneous forces of 100 Newtons [1].

Regarding the cylinder specifically, yes, it is harder to stabilise. The issue is that cylinder/box creates only one contact point, at the cylinder edge. So the cylinder has no other option than to vibrate since the two contact points with each box are always at opposite edges and therefore apply significant alternating torque, which causes vibration. The solution is not in the contact parameters, but in the object design. If you construct your cylinder as a stack of two cylinders, it will be as stable as the others.

Again, I strongly recommend using `simulate` for all of this, you will be able to understand what is happening.

[1] Coin of mass = 0.01 Kg, traveling at 0.1 m/s, hits a window and stops after 10 μs  ->  average force = 1e2 Newtons.

**Reply by stracovskylu** [reply]
Thank you for all your helpful recommendations and feedback. I just have a quick followup questions about the object design: For the case of using a two stacked cylinders instead of one, wouldn't this just make two cylinders which have the same problem of alternating contacts? Perhaps it might reduce the overall alternating torque because two objects would generally ensure two contact points, but I would imagine this still would cause some sort of continuous vibration?
